perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java

```diff
@@ -1670,128 +1776,181 @@ public void saveInformationAboutGroupSynchronization(PerunSession sess, Group gr
+		// Try to find users by login
+		for (Map<String, String> subject: subjects) {
+			String login = subject.get("login");
 			// Skip subjects, which doesn't have login
 			if (login == null || login.isEmpty()) {
-				log.debug("Subject {} doesn't contain attribute login, skipping.", subjectFromLoginSource);
-				skippedMembers.add("MemberEntry:[" + subjectFromLoginSource + "] was skipped because login is missing");
+				log.debug("Subject {} doesn't contain attribute login, skipping.", subject);
+				skippedMembers.add("MemberEntry:[" + subject + "] was skipped because login is missing");
 				continue;
 			}
 
-			//try to find user from perun by login and member extSource (need to use memberSource because loginSource is not saved by synchronization)
 			User user = null;
 			Candidate candidate = null;
 			try {
-				UserExtSource userExtSource = getPerunBl().getUsersManagerBl().getUserExtSourceByExtLogin(sess, memberSource, login);
-				user = getPerunBl().getUsersManagerBl().getUserByUserExtSource(sess, userExtSource);
-				if(!idsOfUsersInGroup.containsKey(user.getId())) {
+				// If login is not in map of actual members, get user and userExtSource from Perun
+				if (!loginsOfMembers.containsKey(login)) {
+					UserExtSource userExtSource = getPerunBl().getUsersManagerBl().getUserExtSourceByExtLogin(sess, memberSource, login);
+					user = getPerunBl().getUsersManagerBl().getUserByUserExtSource(sess, userExtSource);
+
 					candidate = new Candidate(user, userExtSource);
-					//for lightweight synchronization we want to skip all update of attributes
-					candidate.setAttributes(new HashMap<String, String>());
+					// For lightweight synchronization we want to skip all updates of attributes
+					candidate.setAttributes(new HashMap<>());
+				} else {
+					// Get user from richMember, if he is member of group
+					user = loginsOfMembers.get(login).getUser();
 				}
 			} catch (UserExtSourceNotExistsException | UserNotExistsException ex) {
-				//If not find, get more information about him from member extSource
-				List<Map<String, String>> subjectToConvert = Arrays.asList(subjectFromLoginSource);
-				List<Candidate> converetedCandidatesList = convertSubjectsToCandidates(sess, subjectToConvert, memberSource, loginSource, skippedMembers);
-				//Empty means not found (skipped)
-				if(!converetedCandidatesList.isEmpty()) {
-					//We add one subject so we take the one converted candidate
-					candidate = converetedCandidatesList.get(0);
-				}
+				// If not find, get more information about him from member extSource
+				candidate = convertSubjectToCandidate(sess, subject, memberSource, loginSource, skippedMembers);
 			}
 
-			//If user is not null now, we found it so we can use it from perun, in other case he is not in perun at all
+			// If user is not null now, we found it so we can use it from Perun, in other case he is not in Perun at all
 			if(user != null && candidate == null) {
```

Comment:
Because its for the case, when login of this subject is in map "loginsOfMembers", so this member is already in the group. In this case candidate will be null. If his login is not in mentioned map, it tries to find user by UES in Perun. If it found somebody, it creates candidate and without "candidate==null" condition it could end up in this section. And I don't want that.